### PR TITLE
FAI-483: Use constant values for constrained features in CF search

### PR DESCRIPTION
**Pull request by ruivieira**
[FAI-483](https://issues.redhat.com/browse/FAI-483):

The current implementation of CFs requires a search domain for _both_ constrained (fixed) and unconstrained features.

In order to greatly simplify integration with remaining services (and the UI), it should use the original input from the (constrained) feature as a constant value for the CF score calculation, thus not requiring supplying search domains for those inputs.

This PR implements:

- "Fixed" entities for the `CounterfactualEntity` counterparts (e.g. `DoubleEntity` and `FixedDoubleEntity`)
- For fixed entities the search domain is ignored but can be provided optionally (for API compatibility) as a `null` or with `EmptyFeatureDomain.create()`
- Moved the `createEntities` method from `CounterfactualExplainer` to `CounterfactualEntityFactory`

> Many thanks for submitting your Pull Request :heart:! 
> 
> Please make sure that your PR meets the following requirements:
> 
> - [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
> - [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
> - [ ] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
> - [x] Pull Request contains link to the JIRA issue
> - [x] Pull Request contains link to any dependent or related Pull Request
> - [x] Pull Request contains description of the issue
> - [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

**Review by manstis — APPROVED** [review]
Thank-you @ruivieira.

FYI @kelvah until I do https://issues.redhat.com/browse/FAI-486 there is still no validation of the request so you can continue to send what you like.. however when we support structures models and this JIRA is implemented we should probably decide on what the validation should be (for fixed inputs).

**Review by danielezonca — APPROVED** [review]

**Review by r00ta — APPROVED** [review]

**Comment by ruivieira** [comment]
Jenkins retest this please

**Comment by ruivieira** [comment]
Jenkins retest this please

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java [hunk 1]
```diff
@@ -46,6 +53,15 @@ void testIntegerFactory() {
         assertTrue(counterfactualEntity instanceof IntegerEntity);
     }
 
+    @Test
+    void testFixedIntegerFactory() {
+        int value = 5;
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java]: Is value 5 checked somewhere? It seems that it is used in newNumericalFeature but not checked by any assert. Only class of counterfactualEntity is checked.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java [hunk 2]
```diff
@@ -55,6 +71,15 @@ void testDoubleFactory() {
         assertTrue(counterfactualEntity instanceof DoubleEntity);
     }
 
+    @Test
+    void testFixedDoubleFactory() {
+        double value = 5.0;
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java]: Is value 5.0 checked somewhere? It seems that it is used in newNumericalFeature but not checked by any assert. Only class of counterfactualEntity is checked. Also please consider to use "real" double value like 5.5 or something.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java [hunk 3]
```diff
@@ -63,20 +88,36 @@ void testBooleanFactory() {
         assertTrue(counterfactualEntity instanceof BooleanEntity);
     }
 
+    @Test
+    void testFixedBooleanFactory() {
+        final Feature feature = FeatureFactory.newBooleanFeature("bool-feature", false);
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java]: Same as for numerical values.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java [hunk 4]
```diff
@@ -63,20 +88,36 @@ void testBooleanFactory() {
...
         final FeatureDomain domain = CategoricalFeatureDomain.create("foo", "bar");
-        final CounterfactualEntity counterfactualEntity = CounterfactualEntityFactory.from(feature, true, domain);
+        final CounterfactualEntity counterfactualEntity = CounterfactualEntityFactory.from(feature, false, domain);
         assertTrue(counterfactualEntity instanceof CategoricalEntity);
         assertEquals(domain.getCategories(), ((CategoricalEntity) counterfactualEntity).getValueRange());
     }
 
+    @Test
+    void testFixedCategoricalEntity() {
+        final Feature feature = FeatureFactory.newCategoricalFeature("categorical-feature", "foo");
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java]: Same as for numerical value.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java [hunk 5]
```diff
@@ -85,8 +126,42 @@ void testCategoricalFactorySet() {
...
+        featureDomains.add(NumericalFeatureDomain.create(0.0, 1000.0));
+
+        PredictionFeatureDomain featureDomain = new PredictionFeatureDomain(featureDomains);
+
+        PredictionInput input = new PredictionInput(features);
+
+        List<CounterfactualEntity> entities =
+                CounterfactualEntityFactory.createEntities(input, featureDomain, constraints, null);
+
+        assertTrue(entities.get(0) instanceof FixedDoubleEntity);
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java:180]: Can we also check the content of entities?

**Review by jiripetrlik — CHANGES_REQUESTED** [review]
I have some suggestions for improvement of tests.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java (hunk 1 above)
Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualEntityFactoryTest.java]: Hi @jiripetrlik, thanks, good point. Added.

**Review by jiripetrlik — APPROVED** [review]
@ruivieira
Thank you for test enhancements! Looks good to me.

**Comment by ruivieira** [comment]
Jenkins retest this please

**Pull request merged**